persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -1738,6 +1751,181 @@ repsertManySql ent n = putManySql' conflictColumns fields ent n
+
+-- | Copy the field into the database only if the value in the
+-- corresponding record is non-empty, where "empty" means the Monoid
+-- definition for 'mempty'. Useful for 'Text', 'String', 'ByteString', etc.
+--
+-- The resulting 'HandleUpdateCollision' type is useful for the
+-- 'upsertManyWhere' function.
+--
+-- @since  2.12.1.0
+copyUnlessEmpty :: (Monoid.Monoid typ, PersistField typ) => EntityField record typ -> HandleUpdateCollision record
+copyUnlessEmpty field = CopyUnlessEq field Monoid.mempty
+
+-- | Copy the field into the database only if the field is not equal to the
+-- provided value. This is useful to avoid copying weird nullary data into
+-- the database.
+--
+-- The resulting 'HandleUpdateCollision' type is useful for the
+-- 'upsertMany' function.
+--
+-- @since  2.12.1.0
+copyUnlessEq :: PersistField typ => EntityField record typ -> typ -> HandleUpdateCollision record
+copyUnlessEq = CopyUnlessEq
+
+-- | Copy the field directly from the record.
+--
+-- @since 2.12.1.0
+copyField :: PersistField typ => EntityField record typ -> HandleUpdateCollision record
+copyField = CopyField
+
+-- | Postgres specific 'upsertWhere'. This method does the following:
+-- It will insert a record if no matching unique key exists.
+-- If a unique key exists, it will update the relevant field with a user-supplied value, however,
+-- it will only do this update on a user-supplied condition.
+-- For example, here's how this method could be called like such:
+--
+-- upsertWhere record [recordField =. newValue] [recordField /= newValue]
+--
+-- Called thusly, this method will insert a new record (if none exists) OR update a recordField with a new value
+-- assuming the condition in the last block is met.
+-- 
+-- @since 2.12.1.0
+upsertWhere
+  :: ( backend ~ PersistEntityBackend record
+     , PersistEntity record
+     , PersistEntityBackend record ~ SqlBackend
+     , MonadIO m
+     , PersistStore backend
+     , BackendCompatible SqlBackend backend
+     )
+  => record
+  -> [Update record]
+  -> [Filter record]
+  -> ReaderT backend m ()
+upsertWhere record updates filts =
+  upsertManyWhere [record] [] updates filts
+
+-- | Postgres specific 'upsertManyWhere'. This method does the following:
+-- It will insert a record if no matching unique key exists.
+-- If a unique key exists, it will update the relevant field with a user-supplied value, however,
+-- it will only do this update on a user-supplied condition.
+-- For example, here's how this method could be called like such:
+--
+-- upsertManyWhere [record] [recordField =. newValue] [recordField /= newValue]
+--
+-- Called thusly, this method will insert a new record (if none exists) OR update a recordField with a new value
+-- assuming the condition in the last block is met.
+--
+-- -- @since 2.12.1.0
+upsertManyWhere ::
+  forall record backend m.
+  ( backend ~ PersistEntityBackend record,
+    BackendCompatible SqlBackend backend,
+    PersistEntityBackend record ~ SqlBackend,
```

Comment:
I needed to add it to get it compiling tbh; when I get rid of it I get this error 
```
/Users/dmarticus/Personal/github.com/dmarticus/persistent/persistent-postgresql/Database/Persist/Postgresql.hs:1861:5: error:
    • Could not deduce: backend ~ SqlBackend
        arising from a use of ‘mkBulkUpsertQuery’
      from the context: (backend ~ PersistEntityBackend record,
                         BackendCompatible SqlBackend backend, PersistEntity record,
                         MonadIO m)
        bound by the type signature for:
                   upsertManyWhere :: forall record backend (m :: * -> *).
                                      (backend ~ PersistEntityBackend record,
                                       BackendCompatible SqlBackend backend, PersistEntity record,
                                       MonadIO m) =>
                                      [record]
                                      -> [HandleUpdateCollision record]
                                      -> [Update record]
                                      -> [Filter record]
                                      -> ReaderT backend m ()
        at Database/Persist/Postgresql.hs:(1841,1)-(1856,22)
      ‘backend’ is a rigid type variable bound by
        the type signature for:    
          upsertManyWhere :: forall record backend (m :: * -> *).
                             (backend ~ PersistEntityBackend record,
                              BackendCompatible SqlBackend backend, PersistEntity record,
                              MonadIO m) =>
                             [record]
                             -> [HandleUpdateCollision record]
                             -> [Update record]
                             -> [Filter record]
                             -> ReaderT backend m ()
        at Database/Persist/Postgresql.hs:(1841,1)-(1856,22)
    • In the second argument of ‘($)’, namely
        ‘mkBulkUpsertQuery records conn fieldValues updates filters’
      In a stmt of a 'do' block:   
        uncurry rawExecute         
          $ mkBulkUpsertQuery records conn fieldValues updates filters
      In the expression:           
        do conn <- asks projectBackend
           uncurry rawExecute      
             $ mkBulkUpsertQuery records conn fieldValues updates filters
    • Relevant bindings include    
        upsertManyWhere :: [record]
                           -> [HandleUpdateCollision record]
                           -> [Update record]
                           -> [Filter record]
                           -> ReaderT backend m ()
          (bound at Database/Persist/Postgresql.hs:1857:1)
```

Reply:
OK, I see. It's because we're requiring it in `mkBulkUpsertQuery`. But that's also unnecessary. If we delete it from that function, then we can delete it here, too.